proposals/0000-severe-warnings.rst

```diff
@@ -0,0 +1,130 @@
+and upgrades ``missing-methods``,  ``missing-fields`` from ``-Wdefault`` to ``-Wsevere``
+(fixing `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_)
+
+Motivation
+----------
+
+* We can start for the motivation for making ``missing-methods`` an error by default, copying from @phadej’s
+  `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_:
+  
+    Currently omitting method implementations in type-class instances
+    is just a warning, but it is enabled by default.
+    Therefore one could argue that adding a new method to existing
+    type-class is not a breaking change.
+    However, it very much is, but
+
+    *  ecosystem is not forced to catch up in any timely matter
+    *  the breakage is hard to assess
+
+  and (to me even more compelling):
+  
+    FWIW, base-4.18.0.0 is going to have
+
+    Add default implementation of `(<>)` in terms of `sconcat` and `mempty` in terms of `mconcat`.
+    which makes it possible to write::
+
+      instance Semigroup Foo
+      instance Monoid Foo
+
+    and that instance's mempty will loop.
+    Currently GHC will just warn if someone writes such (new) code by accident.
+
+  It seems to me we are doing our users a service if we make that particular foot a little less easy to fire accidentially
+  
+* The motivation for ``-Werror=missing-fields`` follows by dualty – instances are just records for implicitly passed parameters, in a way.
+
+* From the need to have *some* warnings being errors by default follows the need for the suitable infrastructure, hence the
+  ``-Wsevere`` warning group.
+  
+* Given that we have custom warning categories (`x-foo`, `#454 <https://github.com/ghc-proposals/ghc-proposals/pull/454>`_),
+  we anticipate that users of custom warning categories will want a way to add “their” warning group to ``-Wsevere`` as well.
+  
+  Because there is no registry of custom categories, we let the prefix indicate the group. The prefix ``x-`` indicates a custom warning in ``-Wdefault``.
+  This proposal allows the prefix ``xs-` to indicate that this group should be part of the severe group.
+   
+  
+Proposed Change Specification
+-----------------------------
+
+* GHC learns a new warning group flag ``-Wsevere``, next to the existing ``-Wdefault``, ``-Wextra``, `-Wall`` and ``-Wcompat``.
+  See _“5.2. Warnings and sanity-checking” <https://downloads.haskell.org/ghc/latest/docs/users_guide/using-warnings.html>_.
+  
+  It behaves like the other groups in that ``-Wno=severe``, ``-Wwarn=severe`` and ``-Werror=severe`` behave as if the flag
+  had been specified for each warning in the groups.
+  
+* The warning group is initialized with the following warning categories:
+
+  * ``missing-methods``
+  * ``missing-fields``
+
+*  The set of allowed names of custom warning categories (`#454 <https://github.com/ghc-proposals/ghc-proposals/pull/454>`_) is
+   extended by those prefixed with ``xs-*` (for “custom extended – severe”). The flags
```

Comment:
I like this approach, but I'm wondering how the `-Wextended-warnings` group introduced in #541 relates to the new `xs-` prefix. Perhaps `-Wextended-warnings` includes all categories regardless of prefix, but we add a new `-Wsevere-extended-warnings` group that includes only the `xs-` prefixed categories? Thus the default state can easily be described as having `-Wextended-warnings` in `-Wdefault` and `-Wsevere-extended-warnings` in `-Wsevere`.

Do we anticipate wanting to extend this to other warning groups, e.g. `-Wall`?

Reply:
I am not sure if we even need the all-custom-warnings-flag. If we do, adding another for all severe ones is reasonable, I think.

I am not sure if there is demand for off-by-default custom warnings, but I'd there is, `xa-` or something like that would work, would it?